storage/provider/loop.go

```diff
@@ -89,3 +166,72 @@ func (lvs *loopVolumeSource) AttachVolumes(volIds []string, instId []instance.Id
 func (lvs *loopVolumeSource) DetachVolumes(volIds []string, instId []instance.Id) error {
 	panic("not implemented")
 }
+
+func createBlockFile(run RunCommandFn, filePath string, size uint64) error {
+	_, err := run(
+		"dd",
+		"if=/dev/zero",
+		fmt.Sprintf("of=%q", filePath),
+		"bs=1024",
+		fmt.Sprintf("count=%q", size),
+	)
+	return err
+}
+
+func attachToBlockDevice(run RunCommandFn, filePath string) (loopDeviceName string, _ error) {
+	// -f automatically finds the first available loop-device.
+	// --show returns the loop device chosen on stdout.
+	stdout, err := run("losetup", "-f", "--show", filePath)
+	if err != nil && strings.Contains(err.Error(), "could not find any free loop device") {
+		return "", NoLoopDeviceErr
+	}
+	return stdout, err
+}
+
+func detachBlockDevice(run RunCommandFn, devicePath string) error {
+	_, err := run("losetup", "-d", devicePath)
+	return err
+}
+
+func removeBlockFile(run RunCommandFn, filePath string) error {
+	_, err := run("rm", filePath)
+	return err
+}
+
+func createLoopDevice(run RunCommandFn, deviceNum uint) error {
+	_, err := run("mknod", "-m", "660",
+		fmt.Sprintf("/dev/loop%q", deviceNum),
+		"b", "7", fmt.Sprintf("%q", deviceNum),
+	)
+	return err
+}
+
+func findAvailableDeviceNumber(run RunCommandFn) (uint, error) {
+
+	// -a lists all attached loop devices.
+	output, err := run("losetup", "-a")
+	if err != nil {
+		return 0, err
+	} else if output == "" {
+		// Very curious situation to be in. If there are no attached
+		// devices, we should have just attempted to attach to the
+		// next available loop device.
+		return 0, nil
+	}
+
+	// Output will look like: /dev/loop0: [0801]:30551017 (/foo)
+	devNum, err := strconv.Atoi(strings.Split(output, ":")[0][9:])
+	if err != nil {
+		return 0, err
+	}
+
+	return uint(devNum), nil
+}
+
+func providerId(arg storage.VolumeParams) string {
+	uuid, err := utils.NewUUID()
```

Comment:
This needs to be something we can link back to the loop device. Perhaps just compose it from the machine ID and the loop device name. We'd need to pass the machine ID when creating the provider.